### PR TITLE
fix(hcu): support DeepSeek V4 INT8 on gfx936 and bound sparse indexer fallback memory

### PR DESCRIPTION
**Pull request by zhoujie-6**
## 问题
DeepSeek V4 的 compressor 已经写入 packed FP8 indexer 缓存，因此调用 indexer 时可以传入 `k=None`。此前非 gfx938 的 ROCm 分支直接访问 `k.dtype`，并沿用 V3.2 的 BF16 缓存处理路径，无法正确处理这一调用方式和缓存格式。多 token decode 的 PyTorch 实现还存在页面解包方式不一致的问题。

本 MR 增加兼容 V4 packed FP8 缓存的 PyTorch fallback，修正 prefill/decode 计算和不等长批次的权重排列，并通过分块控制临时显存。同时修正 TileLang HIP codegen 探测顺序，以及 ROCm 的 sparse SWA 调度元数据创建路径。

适配目标是 gfx936 上的 DeepSeek-V4-INT8-w8a8模型推理，模型权重为 INT8 W8A8，indexer 缓存为 FP8，两者相互独立。

### 1. V4 packed FP8 缓存读取与平台路由

文件：`vllm_hcu/v1/attention/ops/rocm_aiter_mla_sparse.py`  
函数：`rocm_aiter_sparse_attn_indexer_native()`

对以下条件启用 V4 FP8 fallback：

```python
current_platform.is_rocm() and not on_gfx938() \
    and skip_k_cache_insert and kv_cache.dtype == torch.uint8
```

- 正确处理 `k=None`，避免访问空对象的 `dtype`。
- 使用平台 FP8 dtype 解释缓存，避免将 packed FP8 页面交给 BF16 路径。
- prefill 根据请求边界和 block table 定位物理页面，先 gather 字节，再恢复 FP8 数据和 FP32 scale。
- prefill/decode 分别调用兼容该格式的 PyTorch logits 实现。

缓存按整页组织为：

```text
[整页 FP8 数据字节] + [整页 FP32 scale 字节]
```

该条件覆盖gfx936以及其他非 gfx938 ROCm 平台上的此类 V4 缓存；gfx938 正常执行不会通过此条件进入新增 fallback。

### 2. decode 缓存解包、因果边界和 padding

文件：`vllm_hcu/v1/attention/ops/rocm_aiter_mla_sparse.py`  
函数：`fp8_paged_mqa_logits_torch()`、`rocm_aiter_sparse_attn_indexer_native()`

原实现的单 token 分支按整页读取数据和 scale，多 token 分支却按**最后一维**切分，不能正确解释上述 packed page。本 MR 统一两者的解包方式，并支持一维最终 context 长度和二维逐 query 长度。无效页表项在 gather 前替换为安全页号，超出有效长度的 logits 最终设为 `-inf`。

不等长 decode 中，Q 已按 `decode_lens` 做 padding，但 weights 原先仍保持扁平排列。例如长度 `[1, 3]` 的两个请求，Q 排列为 `[A0, pad, pad]`、`[B0, B1, B2]`，weights 也必须采用相同布局。现在 fallback 在 `requires_padding=True` 时，对 weights 同样调用 `pack_seq_triton`，避免请求之间的权重错位或 reshape 失败。

### 3. prefill scale 广播与设备一致性

文件：`vllm_hcu/v1/attention/ops/rocm_aiter_mla_sparse.py`  
函数：`fp8_mqa_logits_torch()`

多头 score 形状为 `[heads, queries, keys]`，scale 可能是 `[keys]` 或 `[keys, 1]`。本 MR 将 scale 展平并明确沿 key 维应用，避免广播到 query 维或产生形状错误。

创建 mask 等辅助张量时使用 `q.device`，保证与输入设备一致，也支持 CPU 数值测试。

### 4. 分块控制 prefill/decode 临时显存

文件：`vllm_hcu/v1/attention/ops/rocm_aiter_mla_sparse.py`

整批向量化虽然消除了 `.item()` 驱动的 Python 循环，但会完整展开缓存及多头 score。上层按最终 logits 估算的内存预算不足以覆盖这些中间张量。

例如 prefill 的 `heads=64、queries=512、keys=8192` 时，最终 FP32 logits 为 16 MiB，而单个 FP32 score 为 1 GiB，尚未计入类型转换及其他操作的临时分配。

prefill/decode 两条路径使用约 8 MiB 的临时计算预算决定块大小，并逐块写入最终 logits。prefill 使用原位操作并及时释放块内引用，减少中间结果同时存活的数量。

## 测试

gfx936 和 gfx938 平台的精度测试

```bash
vllm serve ./models/DeepSeek-V4-Flash-0731-Channel-INT8-w8a8 \
  --host 0.0.0.0 \
  --port 8002 \
  --served-model-name deepseek-v4-flash-0731 \
  --dtype bfloat16 \
  --kv-cache-dtype fp8 \
  --tensor-parallel-size 8 \
  --max-model-len 4096 \
  --max-num-batched-tokens 512 \
  --max-num-seqs 8 \
  --gpu-memory-utilization 0.80 \
  --block-size 256 \
  --trust-remote-code \
  --tokenizer-mode deepseek_v4 \
  --distributed-executor-backend mp \
  --speculative-config '{"method":"dspark","num_speculative_tokens":5,"draft_sample_method":"probabilistic"}'
```

```bash
evalscope eval \
    --model deepseek-v4-flash-0731 \
    --api-url http://127.0.0.1:8002/v1 \
    --api-key EMPTY \
    --datasets gsm8k \
    --generation-config '{"temperature":0.0,"max_tokens":512,"timeout":1800}' \
    --eval-batch-size 10 \
    --limit 10
```

### CODE REVIEW

File: vllm_hcu/v1/attention/ops/rocm_aiter_mla_sparse.py [hunk 1]
```diff
@@ -1283,7 +1292,20 @@ def rocm_aiter_sparse_attn_indexer_native(
...
             )
-            if not current_platform.is_rocm() or on_gfx938():
+            if v4_fp8_fallback:
+                # Fixed-size GPU gather; sequence boundaries remain device data.
+                page_size = kv_cache.shape[1]
+                pages = kv_cache.view(kv_cache.shape[0], -1)
+                offsets = torch.arange(chunk.total_seq_lens, device=kv_cache.device)
+                seq = torch.searchsorted(chunk.cu_seq_lens[1:].contiguous(), offsets, right=True)
+                local = offsets - chunk.cu_seq_lens[seq]
+                page_ids = chunk.block_table[seq, local // page_size].long()
```
Comment (alexanderbin123) [inline vllm_hcu/v1/attention/ops/rocm_aiter_mla_sparse.py]: [P1] Mask the upper-bound tail before indexing the block table

`chunk.total_seq_lens` is derived from `compressed_seq_lens_cpu`, which the upstream builder explicitly treats as an upper bound for async-spec prefill, while `chunk.cu_seq_lens` is built from the exact device lengths. When replay updates the exact total to a value smaller than that upper bound, every tail offset makes `searchsorted(..., right=True)` return `chunk.num_reqs`; indexing `cu_seq_lens[seq]` still reaches the terminal entry, but this line then indexes row `num_reqs` of a `num_reqs`-row `block_table`, causing an out-of-bounds access/device-side assert. The new graph test only moves the internal boundary while keeping `cu_seq_lens[-1] == total_seq_lens`, so it does not exercise this case. Please mask offsets against `cu_seq_lens[-1]` (using a safe page for the inactive tail) or otherwise make the gather honor the exact total, and add a replay test whose terminal cumulative length shrinks below `total_seq_lens`.

**Review by alexanderbin123 — COMMENTED** [review]
Code review found one blocking correctness issue in the new packed-cache prefill gather.

File: vllm_hcu/v1/attention/ops/rocm_aiter_mla_sparse.py (hunk 1 above)
Comment (zhoujie-6) [reply vllm_hcu/v1/attention/ops/rocm_aiter_mla_sparse.py]: 已在 commit ff2c541 修复：

 - 按 cu_seq_lens[-1] 计算有效 offset mask。在访问 block table 前，将无效尾部的 seq 和 local 置为 0，避免访问 block_table[num_reqs, ...]；随后将尾部的物理页号置为安全页 0。
 - 在写入 k_fp8 / k_scale 前，将无效尾部的数据和 scale 字节清零，避免安全页中的残留值或 NaN 进入下游。
 - 扩展测试，固定 total_seq_lens=7，让精确总长度依次变为 7 → 4 → 0 → 3 → 2 → 7，覆盖尾部缩短、空请求、全部为空及重新增长，并验证有效行数值和尾部清零。CPU 用例直接执行，CUDA 用例使用同一个 Graph replay。

**Review by wics1224 — APPROVED** [review]

File: vllm_hcu/v1/attention/ops/rocm_aiter_mla_sparse.py (hunk 1 above)
Comment (zhoujie-6) [reply vllm_hcu/v1/attention/ops/rocm_aiter_mla_sparse.py]: deepseek-v4 INT8和FP8模型的humaneval精度测试：
```bash
vllm serve /public4/opendas/DL_DATA/llm-models/DeepSeek-V4-Flash-0731-INT8-Channel \
  --host 0.0.0.0 \
  --port 8002 \
  --served-model-name deepseek-v4-flash-0731 \
  --dtype bfloat16 \
  --kv-cache-dtype fp8 \
  --tensor-parallel-size 8 \
  --max-model-len 32768 \
  --max-num-batched-tokens 8192 \
  --max-num-seqs 64 \
  --gpu-memory-utilization 0.80 \
  --block-size 256 \
  --trust-remote-code \
  --tokenizer-mode deepseek_v4 \
  --distributed-executor-backend mp \
  --moe-backend triton \
  --speculative-config '{"method":"dspark","num_speculative_tokens":5,"draft_sample_method":"probabilistic"}'
```

``` bash
evalscope eval \
    --model deepseek-v4-flash-0731 \
    --api-url http://127.0.0.1:8002/v1 \
    --api-key EMPTY \
    --datasets humaneval \
    --generation-config '{"temperature":0.0,"max_tokens":4096,"timeout":1800}' \
    --eval-batch-size 32 \
    --limit 100 
```

| 模型 (Model) | 量化方式 | HumanEval Pass@1 |
| :--- | :--- | :--- |
| dpsk-v4-Flash | INT8 | **0.94** |
| dpsk-v4-Flash | FP8 | **0.92** |

**Comment by zhoujie-6** [comment]
deepseek-v4 INT8和FP8模型的humaneval精度测试：
```bash
vllm serve /public4/opendas/DL_DATA/llm-models/DeepSeek-V4-Flash-0731-INT8-Channel \
  --host 0.0.0.0 \
  --port 8002 \
  --served-model-name deepseek-v4-flash-0731 \
  --dtype bfloat16 \
  --kv-cache-dtype fp8 \
  --tensor-parallel-size 8 \
  --max-model-len 32768 \
  --max-num-batched-tokens 8192 \
  --max-num-seqs 64 \
  --gpu-memory-utilization 0.80 \
  --block-size 256 \
  --trust-remote-code \
  --tokenizer-mode deepseek_v4 \
  --distributed-executor-backend mp \
  --moe-backend triton \
  --speculative-config '{"method":"dspark","num_speculative_tokens":5,"draft_sample_method":"probabilistic"}'
```

``` bash
evalscope eval \
    --model deepseek-v4-flash-0731 \
    --api-url http://127.0.0.1:8002/v1 \
    --api-key EMPTY \
    --datasets humaneval \
    --generation-config '{"temperature":0.0,"max_tokens":4096,"timeout":1800}' \
    --eval-batch-size 32 \
    --limit 100 
```

| 模型 | 量化方式 | 未加 PR | 加 PR 后 | Δ |
| :--- | :--- | :--- | :--- | :--- |
| dpsk-v4-Flash | INT8 | 0.94 | 0.94| 0 |
| dpsk-v4-Flash | FP8  | 0.92 | 0.92| 0 |

**Review by alexanderbin123 — APPROVED** [review]
